File: src/main/scala/com/databricks/labs/validation/Rule.scala
package com.databricks.labs.validation

import com.databricks.labs.validation.utils.Structures.Bounds
import org.apache.log4j.Logger
import org.apache.spark.sql.Column
import org.apache.spark.sql.functions.{array, lit}

/**
 * Definition of a rule
 */
class Rule(
            private val _ruleName: String,
            val inputColumn: Column,
            val ruleType: RuleType.Value
          ) {

  private val logger: Logger = Logger.getLogger(this.getClass)

  private var _boundaries: Bounds = Bounds()
  private var _validExpr: Column = lit(null)
  private var _validNumerics: Column = array(lit(null).cast("double"))
  private var _validStrings: Column = array(lit(null).cast("string"))
  private var _implicitBoolean: Boolean = false
  private var _ignoreCase: Boolean = false
  private var _invertMatch: Boolean = false
  private val inputRuleName: String = setRuleName(_ruleName)
  val inputColumnName: String = inputColumn.expr.toString().replace("'", "")

  override def toString: String = {
    s"""
       |Rule Name: $ruleName
       |Rule Type: $ruleType
       |Rule Is Agg: $isAgg
       |Input Column: ${inputColumn.expr.toString()}
       |Input Column Name: $inputColumnName
       |Boundaries: ${boundaries.lower} - ${boundaries.upper}
       |Valid Numerics: ${validNumerics.expr.toString()}
       |Valid Strings: ${validStrings.expr.toString()}
       |Implicit Bool: ${_implicitBoolean}
       |""".stripMargin
  }

  private def setBoundaries(value: Bounds): this.type = {
    _boundaries = value
    this
  }

  private def setValidNumerics(value: Array[Double]): this.type = {
    _validNumerics = lit(value)
    this
  }

  private def setValidStrings(value: Array[String], ignoreCase: Boolean): this.type = {
    _validStrings = if(ignoreCase) lit(value.map(_.toLowerCase)) else lit(value)
    inputColumn.expr.children.map(_.prettyName)
    this
  }

  private def setValidExpr(value: Column): this.type = {
    _validExpr = lit(value)
    this
  }

  private def setImplicitBool(value: Boolean): this.type = {
    _implicitBoolean = value
    this
  }

  private def setIgnoreCase(value: Boolean): this.type = {
    _ignoreCase = value
    this
  }

  private def setInvertMatch(value: Boolean): this.type = {
    _invertMatch = value
    this
  }

  private def setRuleName(ruleName: String): String = {
    val removedWhitespaceRuleName = ruleName.trim.replaceAll(" ", "_")
    val whitespaceRemovalWarning = s"Converting whitespaces to underscores in Rule's name:\n '$ruleName' --> '$removedWhitespaceRuleName'\n"
    if (_ruleName.contains(" ")) {
      logger.warn(whitespaceRemovalWarning)
      println(whitespaceRemovalWarning)
    }
    val specialCharsPattern = "[^a-zA-z0-9_-]+".r
    val removedSpecialCharsRuleName = removedWhitespaceRuleName.replaceAll("[^a-zA-Z0-9_-]", "_")
    val specialCharacterRemovalWarning = s"Converting special characters to underscores in Rule's name:\n '$removedWhitespaceRuleName' --> '$removedSpecialCharsRuleName'\n"
    if (specialCharsPattern.findAllIn(removedWhitespaceRuleName).toSeq.nonEmpty) {
      logger.warn(specialCharacterRemovalWarning)
      println(specialCharacterRemovalWarning)
    }
    removedSpecialCharsRuleName
  }

  def boundaries: Bounds = _boundaries

  def validNumerics: Column = _validNumerics

  def validStrings: Column = _validStrings

  def validExpr: Column = _validExpr

  def isImplicitBool: Boolean = _implicitBoolean

  def ignoreCase: Boolean = _ignoreCase

  def invertMatch: Boolean = _invertMatch

  def ruleName: String = inputRuleName

  def isAgg: Boolean = {
    inputColumn.expr.prettyName == "aggregateexpression" ||
      inputColumn.expr.children.map(_.prettyName).contains("aggregateexpression")
  }

}

object Rule {

  /**
   * Several apply methods have been created to handle various types of rules and instantiations from the user
   */

  def apply(
             ruleName: String,
             column: Column,
             boundaries: Bounds
           ): Rule = {

    new Rule(ruleName, column, RuleType.ValidateBounds)
      .setBoundaries(boundaries)
  }

  def apply(
             ruleName: String,
             column: Column
           ): Rule = {
      apply(ruleName, column, lit(true))
        .setImplicitBool(true)
  }

  def apply(
             ruleName: String,
             column: Column,
             validExpr: Column
           ): Rule = {

    new Rule(ruleName, column, RuleType.ValidateExpr)
      .setValidExpr(validExpr)
  }

  def apply(
             ruleName: String,
             column: Column,
             validNumerics: Array[Double],
             invertMatch: Boolean
           ): Rule = {

    new Rule(ruleName, column, RuleType.ValidateNumerics)
      .setValidNumerics(validNumerics)
      .setInvertMatch(invertMatch)
  }

  def apply(
             ruleName: String,
             column: Column,
             validNumerics: Array[Double]
           ): Rule = {

    new Rule(ruleName, column, RuleType.ValidateNumerics)
      .setValidNumerics(validNumerics)
      .setInvertMatch(false)
  }

  def apply(
             ruleName: String,
             column: Column,
             validNumerics: Array[Long],
             invertMatch: Boolean
           ): Rule = {

    new Rule(ruleName, column, RuleType.ValidateNumerics)
      .setValidNumerics(validNumerics.map(_.toString.toDouble))
      .setInvertMatch(invertMatch)
  }

  def apply(
             ruleName: String,
             column: Column,
             validNumerics: Array[Long]
           ): Rule = {

    new Rule(ruleName, column, RuleType.ValidateNumerics)
      .setValidNumerics(validNumerics.map(_.toString.toDouble))
      .setInvertMatch(false)
  }

  def apply(
             ruleName: String,
             column: Column,
             validNumerics: Array[Int],
             invertMatch: Boolean
           ): Rule = {

    new Rule(ruleName, column, RuleType.ValidateNumerics)
      .setValidNumerics(validNumerics.map(_.toString.toDouble))
      .setInvertMatch(invertMatch)
  }

  def apply(
             ruleName: String,
             column: Column,
             validNumerics: Array[Int]
           ): Rule = {

    new Rule(ruleName, column, RuleType.ValidateNumerics)
      .setValidNumerics(validNumerics.map(_.toString.toDouble))
      .setInvertMatch(false)
  }

  def apply(
             ruleName: String,
             column: Column,
             validStrings: Array[String],
             ignoreCase: Boolean = false,
             invertMatch: Boolean = false
           ): Rule = {

    new Rule(ruleName, column, RuleType.ValidateStrings)
      .setValidStrings(validStrings, ignoreCase)
      .setIgnoreCase(ignoreCase)
      .setInvertMatch(invertMatch)
  }

}

File: python/databricks/labs/validation/rule.py
import pyspark
from typing import List

from databricks.labs.validation.local_spark_singleton import SparkSingleton
from databricks.labs.validation.rule_type import RuleType
from databricks.labs.validation.structures import Bounds
from databricks.labs.validation.utils.helpers import Helpers


class Rule:
    """
    Definition of a rule
    """

    # TODO: Fix type hint for valid_numerics
    def __init__(self,
                 rule_name: str,
                 column: pyspark.sql.Column,
                 boundaries: Bounds = None,
                 valid_expr: pyspark.sql.Column = None,
                 valid_strings: List[str] = None,
                 valid_numerics=None,
                 ignore_case: bool = False,
                 invert_match: bool = False):

        self._spark = SparkSingleton.get_instance()
        self._column = column
        self._boundaries = boundaries
        self._valid_expr = valid_expr
        self._valid_strings = valid_strings
        self._valid_numerics = valid_numerics
        self._is_implicit_bool = False

        # Determine the Rule type by parsing the input arguments
        if valid_strings is not None and len(valid_strings) > 0:
            j_valid_strings = Helpers.to_java_array(valid_strings, self._spark._sc)
            self._jRule = self._spark._jvm.com.databricks.labs.validation.Rule.apply(rule_name, column._jc,
                                                                                     j_valid_strings,
                                                                                     ignore_case, invert_match)
            self._rule_type = RuleType.ValidateStrings

        elif valid_numerics is not None and len(valid_numerics) > 0:
            j_valid_numerics = Helpers.to_java_array(valid_numerics, self._spark._sc)
            self._jRule = self._spark._jvm.com.databricks.labs.validation.Rule.apply(rule_name,
                                                                                     column._jc,
                                                                                     j_valid_numerics)
            self._rule_type = RuleType.ValidateNumerics
        else:
            self._jRule = self._spark._jvm.com.databricks.labs.validation.Rule.apply(rule_name, column._jc)
            self._is_implicit_bool = True
            self._rule_type = RuleType.ValidateExpr

    def to_string(self):
        return self._jRule.toString()

    def boundaries(self):
        return self._boundaries

    def valid_numerics(self):
        return self._valid_numerics

    def valid_strings(self):
        return self._valid_strings

    def valid_expr(self):
        return self._valid_expr

    def is_implicit_bool(self):
        return self._jRule.implicitBoolean

    def ignore_case(self):
        return self._jRule.ignoreCase

    def invert_match(self):
        return self._jRule.invertMatch

    def rule_name(self):
        return self._jRule.ruleName

    def is_agg(self):
        return self._jRule.isAgg

    def input_column_name(self):
        return self._jRule.inputColumnName

    def rule_type(self):
        return self._rule_type

    def to_java(self):
        return self._jRule
